### PR TITLE
Challenge/longest substring

### PR DESCRIPTION
**Pull request by imwaymaran**
---
## about: A function that finds the length of the longest substring without repeating characters.
---

## Behavior
The function determines the length of the longest substring in a given string `s` that does not contain any repeating characters.

* **Input**:  
  - The function accepts a single string `s`.  
  - The input must satisfy the following constraint:  
    `0 <= len(s) <= 5 * 10^4`.  
  - The string can consist of English letters (both uppercase and lowercase), digits, symbols, and spaces.

* **Output**:  
  - The function returns an integer representing the length of the longest substring without repeating characters.  
  - If the input string is empty, the function returns `0`.

* **Error Handling**:  
  - If the input is not a string, the function raises an `AssertionError` with the message "Given input must be a string."  
  - If the input string length is outside the allowed range (0 to 50,000 characters), the function raises an `AssertionError` with the message "Input string length must be between 0 and 50,000 characters."

The function uses a **sliding window technique** to efficiently track unique characters in a substring, adjusting the window size dynamically as it iterates through the input string. Defensive assertions ensure that the input is valid before processing.

---
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/longest_substring.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    # Assert that the input is a string
+    assert isinstance(s, str), "Given input must be a string."
+
+    # Assert that the string length is within the allowed limit
+    assert (
+        0 <= len(s) <= 5 * 10**4
+    ), "Input string length must be between 0 and 50,000 characters."
+
+    # Use a sliding window strategy to track unique characters in the current substring
```
Comment (doctorbanu) [inline solutions/longest_substring.py:48]: Adding a brief explanation of the sliding window strategy in the function might be more beginner-friendly and help those who are less familiar with the approach. (Like me😅)

File: solutions/tests/test_longest_substring.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    def test_large_input(self):
+        """Test with a long input string, expecting output equal to unique characters."""
+        large_input = (
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~ "
+            * 10  # Repeating the pattern to make the string longer
+        )
+        expected_output = len(set(large_input))  # Unique characters in the input
+        self.assertEqual(longest_substring(large_input), expected_output)
+
+    # Edge cases
```
Comment (doctorbanu) [inline solutions/tests/test_longest_substring.py:55]: It could be beneficial to include a couple of additional edge case tests that cover a variety of character combinations. This might include strings with mixed uppercase and lowercase letters, numbers combined with letters or spaces, special symbols mixed with numbers or alphabetic characters, and spaces interspersed with any of these elements, etc. (Of course not all of them but some of them, I hope you got the idea.) I think, this type of combinations could help ensure the function handles different types of input more comprehensively.

**Review by doctorbanu — APPROVED** [review]
Overall, the code and tests are exceptional—clear, structured, and thorough. The implementation is solid, and the test suite covers a wide range of scenarios, including standard, performance, edge, and defensive cases. 

I have 2 very minor suggestions, but your files are already excellent. Seriously, I couldn’t really find any extra recommendations. This is a fantastic work, and I’m approving your pull request. But before merging, I will wait for your feedback on my comments just in case. Well done Gennadii!

File: solutions/longest_substring.py (hunk 1 above)
Comment (imwaymaran) [reply solutions/longest_substring.py:48]: Honestly, I wasn’t familiar with the sliding window technique before this challenge either. I intentionally kept the comments minimal because I feel like they should focus on the strategy, not the implementation details. But I get how a brief explanation could be helpful for those who aren’t familiar with it. Thanks for pointing that out!

File: solutions/tests/test_longest_substring.py (hunk 2 above)
Comment (imwaymaran) [reply solutions/tests/test_longest_substring.py:55]: Thanks for the suggestion! After looking through the tests, I think the current ones already cover a lot of different cases:
	•	Standard test cases: We have examples with both repeating and unique characters, like "abcabcbb" and "bbbbb".
	•	Performance tests: We’re testing large strings with a mix of letters, digits, and symbols to check how the function handles big inputs.
	•	Edge cases: We cover cases like empty strings, single characters, repeating patterns, and strings that hit the 50,000 character limit.
	•	Defensive tests: There are tests for invalid inputs like integers, lists, and None.

The tests you suggested, with mixed case letters, spaces, and symbols, are probably already covered in our test_large_input and test_repeating_pattern tests, which include a variety of characters. So I believe we’re already testing a pretty wide range of possibilities, and adding more might not be necessary.

Thanks again for the input though! I hope this helps explain why I think we’re good on test coverage.

Comment (imwaymaran) [reply solutions/tests/test_longest_substring.py:55]: Thanks for the suggestion! After reviewing the tests, I believe the current ones already cover a wide variety of cases:
	•	Standard test cases: I’ve got examples with both repeating and unique characters, like “abcabcbb” and “bbbbb”.
	•	Performance tests: I’m testing large strings with a mix of letters, digits, and symbols to see how the function handles big inputs.
	•	Edge cases: I cover cases such as empty strings, single characters, repeating patterns, and strings that reach the 50,000 character limit.
	•	Defensive tests: There are tests for invalid inputs like integers, lists, and None.

The types of tests you suggested, with mixed-case letters, spaces, and symbols, are already likely covered in the test_large_input and test_repeating_pattern tests, which include a variety of characters. So, I think I’m already testing a broad range of scenarios, and adding more might not be necessary.

Thanks again for the input! I hope this helps clarify why I believe I’m good with the current test coverage.

Comment (doctorbanu) [reply solutions/tests/test_longest_substring.py:55]: Yes, definitely. Your tests are very well done and cover a broad spectrum. That was just an extra idea, an additional recommendation. Since you have already done everything flawlessly, I really couldn’t find anything else to add. I just wanted to share it when something like this came to my mind.
